File: solutions/free-pascal/prime-factors/1/PrimeFactors.pas
unit PrimeFactors;

{$mode ObjFPC}{$H+}

interface

type
  TIntArray = Array Of Integer;

function factors(
  const value   : integer
) : TIntArray;

implementation

uses SysUtils;

function factors(
  const value   : integer
) : TIntArray;
var
  num, k : integer;
begin
  num := value;
  result := [];
  for k:=2 to Round(sqrt(num)) do
  begin
    while num mod k = 0 do
    begin
      insert(k, result, length(result));      
      num := num div k;
    end;
    if num=1 then
      break;
  end;
if num>1 then
  insert(num, result, length(result)); 
end;

end.

File: solutions/cpp/prime-factors/3/prime_factors.cpp
#include "prime_factors.h"
#include <vector>
using namespace std;

namespace prime_factors {
    vector<long long> of(long int num){
        vector<long long> factors{};
        for(auto k=2; k*k<=num ;k++)
            while(!(num % k)){
                factors.push_back(k);
                num /= k;
                }
       if (num>1) 
            factors.push_back(num);
        return factors;
        }
}  // namespace prime_factors
